ENH: Add vtkMRMLTextNode and "Texts" module

Sunderlandkyl opened the pull request:
vtkMRMLTextNode and vtkMRMLTextStorageNode migrated to Slicer core from SlicerOpenIGTLink.

The Texts module adds a new designer plugin widget: "qMRMLTextWidget".
qMRMLText provides an text box with properties for readOnly, and continuousUpdate.

![image](https://user-images.githubusercontent.com/9222709/58732525-c8a8ab00-83bf-11e9-95a0-96f70b2e81a4.png)

Sunderlandkyl commented:
- Added vtkMRMLTextNode and vtkMRMLTextStorageNode tests.
- Added warning messages when saving/cancelling an edit in qMRMLTextWidget could overwrite/discard data:
![image](https://user-images.githubusercontent.com/9222709/58892918-20475f00-86bd-11e9-86e4-62c2c85028d7.png) 
![image](https://user-images.githubusercontent.com/9222709/58892930-23dae600-86bd-11e9-8bb1-52948c5afb26.png)

jcfr commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,131 @@
+  bool readOnly();
+
+  /// Returns true if the text editor is continuously updated from the text node and vice versa
+  bool continuousUpdate();
```
What do you think of using `autoUpdate` ?  This would be consistent with the  `updateWidget` and `updateFinished` method names

Sunderlandkyl replied:
Sounds good! autoUpdate() is much better than continuousUpdate().

Sunderlandkyl replied:
Done!

jcfr commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,131 @@
+  void onTextChanged();
+
+  /// Method invoked when the "Edit" button is clicked
+  void onEditClicked();
```
onEditClicked -> onEditButtonClicked ? Same comment for Cancel and Save

That said, I wonder if it makes sense to have these signals. Having signal like `updateRequested` and `updateCanceled` would be more generic  and would also nicely complement the existing `updateFinished` signal

Sunderlandkyl replied:
Changed methods to XButtonClicked(). Added new signals (updateRequested, editingChanged(bool)).

jcfr commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,131 @@
+  /// This signal is emitted if updates to the widget have finished.
+  void updateFinished();
+
+  /// This signal is emitted if the node changes
```
```
\sa setMRMLNode, mrmlNode()
```

Sunderlandkyl replied:
Done!

jcfr commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,131 @@
+  Q_PROPERTY(bool readOnly READ readOnly WRITE setReadOnly)
+  Q_PROPERTY(bool continuousUpdate READ continuousUpdate WRITE setContinuousUpdate)
+
+  /// Get the table node the columns are edited of.
```
Comment should be updated

Sunderlandkyl replied:
Done!

jcfr commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,131 @@
+  /// This signal is emitted if the node changes
+  void mrmlNodeChanged(vtkMRMLNode*);
+
+  /// This signal is emitted if the read only propery is changed
```
propery -> property

Sunderlandkyl replied:
Done!

jcfr commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,131 @@
+  void mrmlNodeChanged(vtkMRMLNode*);
+
+  /// This signal is emitted if the read only propery is changed
+  void readOnlyChanged(bool);
```
Instead of having a signal in the widget, should we instead have a a property in the text node ?

Sunderlandkyl replied:
I think that they would serve two different purposes, although having a readOnly property in the text node does seem like a useful feature.

- Read only in the qMRMLTextWidget allows text node content display without allowing editing (Whether or not the node is read only).

Sunderlandkyl replied:
Left ReadOnly out of the node for now.

Re a discussion with @lassoan, If it was implemented, it would probably be like the Locked parameter in markup and table nodes.

jcfr commented on Libs/MRML/Core/vtkMRMLTextNode.h:
```diff
@@ -0,0 +1,104 @@
+    ENCODING_US_ASCII = 3,
+    ENCODING_ISO_8859_1 = 4,
+    ENCODING_LATIN1 = ENCODING_ISO_8859_1 // alias
+                      // see other codes at http://www.iana.org/assignments/character-sets/character-sets.xhtml
```
It would be nice to also support UTF8 and add test for it.

Sunderlandkyl replied:
Added a test for UTF-8, and decided to just use the VTK_ENCODING_X enum to represent encoding.

jcfr reviewed (changes requested):
This is great ! Thanks for working on this. 

Having first class support for text base document like json, xml, txt is very nice.

I wonder if this should be used as a base for other node stored as text or json (e.g markup)

Sunderlandkyl commented on Libs/MRML/Core/vtkMRMLTextNode.cxx:
```diff
@@ -0,0 +1,165 @@
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLTextNode::SetText(const char* text)
```
Changed const* char to std::string.

Sunderlandkyl commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,131 @@
+  Q_INVOKABLE vtkMRMLNode* mrmlNode() const;
+
+  /// Returns true if the text editor is read only
+  bool readOnly();
```
Done!

lassoan commented on Libs/MRML/Core/vtkMRMLTextNode.cxx:
```diff
@@ -0,0 +1,148 @@
+vtkMRMLTextNode::vtkMRMLTextNode()
+  : Text("")
+  , Encoding(VTK_ENCODING_US_ASCII)
+  , ForceStorageNode(false)
```
It would be better to use the variable name ForceCreateStorageNode with enum value: CreateStorageNodeAuto, CreateStorageNodeAlways, CreateStorageNodeNever (to make the behavior more predictable).

lassoan commented on Libs/MRML/Core/vtkMRMLTextNode.cxx:
```diff
@@ -0,0 +1,148 @@
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLTextNode::SetText(std::string text, int encoding/*VTK_ENCODING_NONE*/)
```
Performance could be slightly better with const std::string &text (which might matter if the string is long and updates are frequent). If you don't see any drawback then use const ref instead.

lassoan commented on Libs/MRML/Core/vtkMRMLTextNode.cxx:
```diff
@@ -0,0 +1,148 @@
+  if (!this->ForceStorageNode)
+    {
+    int length = this->Text.length();
+    if (length < LENGTH_BEFORE_STORAGE_NODE)
```
MAX_STRING_LENGTH_FOR_SAVE_WITHOUT_STORAGE_NODE

lassoan commented on Libs/MRML/Core/vtkMRMLTextNode.h:
```diff
@@ -0,0 +1,96 @@
+  virtual const char* GetNodeTagName() override { return "Text"; };
+
+  /// Set text node contents
+  void SetText(const std::string text, int encoding=VTK_ENCODING_NONE);
```
It is interesting that this compiled (const is in the header but not in the implementation). Anyway, it probably should be const std::string&.

lassoan commented on Libs/MRML/Core/vtkMRMLTextNode.h:
```diff
@@ -0,0 +1,96 @@
+  /// For character encoding, please refer IANA Character Sets
+  /// (http://www.iana.org/assignments/character-sets/character-sets.xhtml)
+  /// Default is VTK_ENCODING_US_ASCII
+  vtkSetClampMacro(Encoding, int, VTK_ENCODING_NONE, VTK_ENCODING_UNKNOWN);
```
Probably we should invoke TextModifiedEvent even if encoding was changed.

lassoan commented on Libs/MRML/Core/vtkMRMLTextNode.h:
```diff
@@ -0,0 +1,96 @@
+
+  enum
+  {
+    TextModifiedEvent = 51000,
```
Document when this event is invoked. Also document if Modified event is not invoked for certain Set methods.

lassoan commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx:
```diff
@@ -0,0 +1,382 @@
+  this->qvtkReconnect(d->CurrentTextNode, node, vtkMRMLTextNode::TextModifiedEvent, this, SLOT(onTextNodeModified()));
+  d->CurrentTextNode = node;
+  d->Editing = false;
+  this->updateRequested();
```
Direction of update is not clear. Use updateWidgetFromMRML (updateWidgetFromMRMLRequested).

lassoan commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx:
```diff
@@ -0,0 +1,382 @@
+void qMRMLTextWidget::setReadOnly(bool readOnly)
+{
+  Q_D(qMRMLTextWidget);
+  if (this->ReadOnly == readOnly)
```
What happens if you set read-only while the widget is being edited? Do you save the results? (probably should)

lassoan commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx:
```diff
@@ -0,0 +1,382 @@
+
+  if (d->TextEditModified)
+    {
+    QMessageBox msgBox;
```
Proper QMessageBox deleting is tricky. You may need to make it auto-close if you don't destroy explicitly. Search for other similar messageboxes and copy their implementation.

Sunderlandkyl replied:
This is how QMessageBox and ctkMessageBox are handled in qMRMLSceneViewMenu and qMRMLAnnotationTreeView.

Is there another style of implementation that would be better?

lassoan commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.cxx:
```diff
@@ -0,0 +1,382 @@
+  if (d->TextNodeModified)
+    {
+    QMessageBox msgBox;
+    msgBox.setText("The original text has been modified.");
```
Mention that the text has been modified _since editing was started_.

lassoan commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,139 @@
+  /// Returns true if the text editor is read only
+  bool isReadOnly();
+
+  /// Returns true if the text editor is continuously updated from the text node and vice versa
```
Add a note on how auto-update works while editing is in progress.

lassoan commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,139 @@
+  Q_INVOKABLE vtkMRMLTextNode* mrmlTextNode() const;
+  Q_INVOKABLE vtkMRMLNode* mrmlNode() const;
+
+  /// Returns true if the text editor is read only
```
Add some more info about this. How the appearance and behavior is different if read-only or not.

lassoan commented on Modules/Loadable/Texts/Widgets/qMRMLTextWidget.h:
```diff
@@ -0,0 +1,139 @@
+
+signals:
+  /// This signal is emitted if updates to the widget have been requested.
+  void updateRequested();
```
rename to update...From...Requested

lassoan commented on Modules/Loadable/Texts/qSlicerTextsModule.cxx:
```diff
@@ -0,0 +1,129 @@
+QStringList qSlicerTextsModule::categories() const
+{
+  return QStringList()
+    << "";
```
Put it in the same category as Tables (Informatics).

lassoan reviewed (commented):
Very nice, thank you. I just added a couple of small comments that should be mostly trivial to address.

Sunderlandkyl commented:
I made most of the changes (With the exception of the QMessageBox change), and added some more lines in the test for the new storage node options (CreateStorageNodeAlways, CreateStorageNodeAuto, CreateStorageNodeNever).

Sunderlandkyl commented:
Message boxes removed in favor of a warning label under the save button:
![image](https://user-images.githubusercontent.com/9222709/59438966-6ccc2200-8dc2-11e9-9edd-6f2222c640f0.png)

lassoan commented:
Great, thank you. Maybe we could make it more subtle? For example, add a warning icon to the "Save" button and add the "The original text has been modified..." text to the button's tooltip. This way, we would not need to change the size of the widget and the warning icon may be more visible then the text below.

Sunderlandkyl commented:
Changed to icon+tooltip on save button:

![image](https://user-images.githubusercontent.com/9222709/59453350-b2e2af00-8ddd-11e9-965c-cc43d405aa85.png)

lassoan commented:
Thank you, this looks good. I'll merge this next week after the CARS tutorial is finished, just to make sure we don't introduce any regressions in the last minute.

Sunderlandkyl committed:
ENH: Add vtkMRMLTextNode and "Texts" module

vtkMRMLTextNode and vtkMRMLTextStorageNode migrated to Slicer core from SlicerOpenIGTLink.

The Texts module adds a new designer plugin widget: "qMRMLTextWidget".
qMRMLText provides an text box with properties for readOnly, and continuousUpdate.

lassoan committed:
ENH: Text widget enhancements

Rename auto-update to auto-save.
Auto-select existing text node when opening Texts module.
Fixed crash when attempting to edit node after scene close.
Exposed text edit widget and editing saving/cancelling.
Made editing property accessible from Python.

lassoan commented:
I've reviewed and tested this and pushed some fixes/improvements. There are just a few things that would be nice to address before merging:
- [ ] Clear widget contents when scene is closed.
- [ ] Add word wrap option (Boolean flag in the text edit widget and a checkbox in the module widget's advanced section)
- [ ] Double-check that Qt designer plugin works (for me the text edit widget did not show up in QtDesigner, but maybe I just did not rebuild enough)

Sunderlandkyl committed:
ENH: Add Texts module subject hierarchy plugin

Features:
- Add word wrap option to widget
- Add subject hierarchy plugin
- Add function vtkMRMLTextNode::GetEncodingAsString()

Fixes:
- Widget cleared when scene is closed
- Only remove last file extensions
- SetEncoding issue where every encoding was set to "Unknown"

Sunderlandkyl commented:
Text nodes in SH currently show node contents in tooltip:
![image](https://user-images.githubusercontent.com/9222709/59879252-3ec57f80-9378-11e9-9f39-96bb5e3c5235.png)

"Edit text" in SH context menu:
![image](https://user-images.githubusercontent.com/9222709/59879367-864c0b80-9378-11e9-9a83-ca4f1d1fae2d.png)

lassoan commented:
Looks good!
What does "Edit text" do? If the same as "Edit properties..." (open Texts module with the text node selected) then it is probably redundant.

Sunderlandkyl commented:
Yeah it does. Didn't even notice that "Edit properties..." was there.

Sunderlandkyl commented:
Removed "Edit text" and added encoding info to the SH tooltip.

lassoan commented:
@Sunderlandkyl 

> added encoding info to the SH tooltip.

I'm not sure if this is needed. Encoding is only needed to interpret the byte stream and convert it to displayable string. If we already display the string in the tooltip it does not really matter what encoding was used (especially not for users). If the text is empty or very short then displaying encoding information might be distracting.

Let me know if you could address the other things (scene close, etc.) and it is ready for testing. Thank you!

@jcfr 

> I wonder if this should be used as a base for other node stored as text or json (e.g markup)

This is a good idea, but we need to think about it a bit more. This pull request already contains many changes, so I would merge this and get back to this idea later.

Sunderlandkyl commented:
Sorry, the other issues have been addressed.

What do you think the ideal tooltip should be?

1.
```
Text (Encoding: ASCII)
// Node contents
```
2.
```
Text
// Node contents
```
3.
```
// Node contents
```

1 and 2 are more consistent with other node types, while is much simpler.

lassoan commented:
I like both 1 and 3 (1 is more consistent with other nodes; 3 is more to the point). Let's just keep the current implementation then. We can tune later based on real-world experience. I'll merge this now.

lassoan commented:
Merged in r28328. Thanks a lot for the great work!

You can update SlicerOpenIGTLink now to not use its own text node in Slicer-4.11.

Sunderlandkyl commented:
Great, thanks Andras!

jcfr commented:
Currently working on GitHub transtion, data migration, ... I added a note to the authorship page to add you back as author with Andras as co-author and committer. See https://www.slicer.org/wiki/Documentation/Labs/TransitionToGit#Authorship
